### PR TITLE
Converge common msbuild properties and targets

### PR DESCRIPTION
**Pull request by ViktorHofer**
Move common msbuild properties and targets into the repo root.

@trylek I don't really understand the difference between dir.common.props/targets and Directory.Build.props/targets. I think we should clean this up so that only the latter exists and is imported everywhere.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -49,6 +57,49 @@
...
+    <SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
+
+    <!-- Resource naming bug: https://github.com/microsoft/msbuild/issues/4740 -->
+    <EmbeddedResourceUseDependentUponConvention>false</EmbeddedResourceUseDependentUponConvention>
+  </PropertyGroup>
+
+  <!-- Analyzers -->
+  <PropertyGroup>
+    <CodeAnalysisRuleset>$(LibrariesProjectRoot)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <EnablePinvokeUWPAnalyzer>false</EnablePinvokeUWPAnalyzer>
```
Comment (jkotas) [inline Directory.Build.props]: We have dropped UWP target. Is this property still needed? And if it is still needed, is it relevant to anything outside libraries?

File: Directory.Build.props [hunk 2]
```diff
@@ -49,6 +57,49 @@
...
+  </PropertyGroup>
+
+  <!-- Set some shortcuts for more terse conditions in project files -->
```
Comment (jkotas) [inline Directory.Build.props]: Is this relevant to anything outside libraries? I would expect everything outside libraries targets netcoreapp.

File: Directory.Build.targets [hunk 3]
```diff
@@ -1,5 +1,33 @@
...
+    <!--
+    For non-SDK projects that import this file and then import Microsoft.Common.targets,
+    tell Microsoft.Common.targets not to import Directory.Build.targets again
+    -->
+    <ImportDirectoryBuildTargets>false</ImportDirectoryBuildTargets>
+  </PropertyGroup>
+
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(SkipImportArcadeSdkFromRoot)' != 'true'" />
 
+  <!-- Adds Nullable annotation attributes to netstandard <= 2.0 builds -->
```
Comment (jkotas) [inline Directory.Build.targets]: Is this relevant to anything outside libraries?

File: src/libraries/Directory.Build.props [hunk 4]
```diff
@@ -119,16 +97,16 @@
   <PropertyGroup>
     <!-- By default make all libraries to be AnyCPU but individual projects can override it if they need to -->
     <Platform>AnyCPU</Platform>
-    <OutputType>Library</OutputType>
```
Comment (jkotas) [inline src/libraries/Directory.Build.props]: Just curious - where did this go?

File: Directory.Build.props [hunk 5]
```diff
@@ -49,6 +57,49 @@
...
+    <!-- Suppress preview message as we are usually using preview SDK versions. -->
+    <SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
+
+    <!-- Resource naming bug: https://github.com/microsoft/msbuild/issues/4740 -->
+    <EmbeddedResourceUseDependentUponConvention>false</EmbeddedResourceUseDependentUponConvention>
+  </PropertyGroup>
+
+  <!-- Analyzers -->
+  <PropertyGroup>
+    <CodeAnalysisRuleset>$(LibrariesProjectRoot)CodeAnalysis.ruleset</CodeAnalysisRuleset>
```
Comment (jkotas) [inline Directory.Build.props]: Majority of the C# code outside libraries is not polished to follow all these codeanalysis rules. Should this stay specific to libraries?

File: src/libraries/Directory.Build.props (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props]: this is the default, no need to be explicit here.

File: Directory.Build.props (hunk 5 above)
Comment (ViktorHofer) [reply Directory.Build.props]: That file is used by CoreLib, which lives outside libraries, as well. Extracting it into common makes most sense.

File: Directory.Build.targets (hunk 3 above)
Comment (ViktorHofer) [reply Directory.Build.targets]: Any library outside libraries that wants to enable nullable annotations which doesn't target .NET Core >= 3.0 benefits from this. I think there are cases in the other partitions where this might be useful.

File: Directory.Build.props (hunk 2 above)
Comment (ViktorHofer) [reply Directory.Build.props]: There is at least tools-local which multi-targets and I'm sure there are other cases. That said, I don't have a strong opinion.

File: Directory.Build.props (hunk 1 above)
Comment (ViktorHofer) [reply Directory.Build.props]: This property affects every project that references `Microsoft.DotNet.CodeAnalysis`: https://github.com/dotnet/arcade/blob/7584330809c5463245e24f0f86ac2531a36dacd5/src/Microsoft.DotNet.CodeAnalysis/build/Microsoft.DotNet.CodeAnalysis.targets#L6. The right fix here is to flip the bit there. I will submit a PR.

Comment (ViktorHofer) [reply Directory.Build.props]: https://github.com/dotnet/arcade/pull/4446

File: Directory.Build.targets (hunk 3 above)
Comment (jkotas) [reply Directory.Build.targets]: What are those cases? I expect that we will move all libraries that ship as netstandard targeting NuGet packages to live under libraries.

File: Directory.Build.props (hunk 5 above)
Comment (jkotas) [reply Directory.Build.props]: CoreLib is referencing all its source files from under libraries, so it can reference the codeanalysis ruleset from under libraries as well.

Comment (ViktorHofer) [reply Directory.Build.props]: > CoreLib is referencing all its source files from under libraries

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj#L137-L412?

Comment (jkotas) [reply Directory.Build.props]: I meant to say`CoreLib is referencing a lot of of its source files from under libraries.

Comment (ViktorHofer) [reply Directory.Build.props]: If the plan is to move all managed projects to libraries than I agree with what you are saying but I don't think that's the plan? Also this property doesn't trigger thousands of errors/warnings but just sets the path to the rulset file in case a project references the analyzers. cc @stephentoub for dicussion.

Comment (jkotas) [reply Directory.Build.props]: We are going to have many managed projects outside libraries.

Comment (jkotas) [reply Directory.Build.props]: > just sets the path to the rulset file in case a project references the analyzers

Ok, if this is a link to file with reasonable set of defaults, this can be set globally. Should the file move to a more global place too then?

Comment (jkotas) [reply Directory.Build.props]: Also, what one has to do to enable use of the default ruleset? Should we also have the references to the analyzer packages, etc. here so that one can enable use of the default ruleset via a single property?

Comment (ViktorHofer) [reply Directory.Build.props]: > Also, what one has to do to enable use of the default ruleset? Should we also have the references to the analyzer packages, etc. here so that one can enable use of the default ruleset via a single property

AFAIK that's @stephentoub's plan.

> Should the file move to a more global place too then?

I proposed that as well but @stephentoub was against it. I think if we move all logic into common we should move that file as well.

Comment (stephentoub) [reply Directory.Build.props]: I wasn't against it. I just didn't want to expand that PR to consolidate all the logic that's currently spread across several places. I'm happy for all analysis support for managed code in the repo to be consolidated to one top-level location in the repo.

Comment (ViktorHofer) [reply Directory.Build.props]: I was referring to this comment: https://github.com/dotnet/runtime/pull/59#issuecomment-554401215. Anyways as you just agreed with moving everything analyzer related into root, I'm happy to do so :)

Comment (stephentoub) [reply Directory.Build.props]: Right. If we're only analyzing src in libraries (plus a few corelib files), it shouldn't be at the root. But this discussion suggests we're talking about enabling it for lots of stuff outside of libraries. Also, I think it should be consolidated, but not at the root of the repo. Isn't it better in eng?

Comment (ViktorHofer) [reply Directory.Build.props]: The [analyzers.props](https://github.com/dotnet/runtime/blob/master/eng/analyzers.props) file already lives in eng. I think it makes most sense to move the property that points to the ruleset file there too. We can move the rulset file there as well, I don't have a strong opinion on that.

Comment (stephentoub) [reply Directory.Build.props]: Sounds like the ruleset reference should just be added to that file and the ruleset put next to it.

Comment (ViktorHofer) [reply Directory.Build.props]: Perfect.

File: Directory.Build.props (hunk 2 above)
Comment (ViktorHofer) [reply Directory.Build.props]: I had to move those back anyway as they rely on configurations set in libraries which happen at a later point.

File: Directory.Build.targets (hunk 3 above)
Comment (ViktorHofer) [reply Directory.Build.targets]: That's a good argument. Will move it back. I was thinking about enabling this scenario just entirely in the repo but if there's no use case for it then this might be over-engineering.

**Review by jkotas — APPROVED** [review]

**Review by trylek — APPROVED** [review]
Looks good, thanks for cleaning this up. I'll take a look at the extra CoreCLR scripts.

**Comment by ViktorHofer** [comment]
Had to force push without changing anything (Arcade update wasn't picked up in a retry).

File: src/coreclr/src/System.Private.CoreLib/ILLink.targets [hunk 6]
```diff
@@ -1,146 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/ILLink.targets]: There is important logic in this file that auto-generates the ILLinkXml file based on the unmanaged runtime dependencies. It is not ok to just delete it.

If you want to refactor this or clean it up, it would be best done as separate PR.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/ILLink.targets]: I diffed the contents and went through git history. It's a duplicate of eng/illink.targets.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/ILLink.targets]: The special thing that you are referring to is probably https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/CreateRuntimeRootILLinkDescriptorFile.targets which will stay as is.

File: eng/illink.targets [hunk 7]
```diff
@@ -24,7 +24,7 @@
     <ILLinkRewritePDBs Condition="'$(ILLinkRewritePDBs)' == '' and '$(DebugSymbols)' != 'false'">true</ILLinkRewritePDBs>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(ILLinkTrimXml)' != ''">
+  <ItemGroup Condition="Exists('$(ILLinkTrimXml)')">
```
Comment (jkotas) [inline eng/illink.targets]: The exists condition like this just make for hard to debug problems. If one specifies ILLinkTrimXml file, it should better exist.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 8]
```diff
@@ -20,6 +18,10 @@
     <OutputPath>$(BinDir)IL/</OutputPath>
     <Configurations>Debug;Release;Checked</Configurations>
     <Platforms>x64;x86;arm;arm64</Platforms>
+
+    <ILLinkClearInitLocals>true</ILLinkClearInitLocals>
+    <ILLinkTrimAssembly>true</ILLinkTrimAssembly>
+    <ILLinkTrimXml>unused</ILLinkTrimXml>
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Should this rather point to the auto-generated ILLinkTrimXml under obj directory?

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 9]
```diff
@@ -96,10 +98,6 @@
 
   <!-- Assembly attributes -->
   <PropertyGroup>
-    <AssemblyName>System.Private.CoreLib</AssemblyName>
-    <AssemblyVersion>5.0.0.0</AssemblyVersion>
-    <ExcludeAssemblyInfoPartialFile>true</ExcludeAssemblyInfoPartialFile>
-    <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Why do we need to generate the target framework atrribute in CoreLib? It is just an extra cruft that nobody cares about.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: We generate the attribute for all assemblies in libraries, even for Private ones like System.Private.Xml. Mainly I did this to follow-up on your comment here: https://github.com/dotnet/corefx/pull/32680#issuecomment-429478625

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 8 above)
Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: You are absolutely right. Pushed a commit.

File: eng/illink.targets (hunk 7 above)
Comment (ViktorHofer) [reply eng/illink.targets]: reverted

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 9 above)
Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Ah, ok - I forgot about that discussion. As you can tell, I still wonder why we are generating this cruft.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: > I still wonder why we are generating this cruft for inbox assemblies.

It's funny because in the past you mentioned that we should enable it for all - even the inbox ones. I don't have a strong opinion here but I would like to enable this for the sake of consistency. Also I prefer not opting out of the default sdk behavior if there isn't a strong reason.

Also as a side note, I noticed partners taking a dependency on the attribute, ie vstest for test assemblies: https://github.com/microsoft/vstest/blob/7b6248203164f8ea821f6795632bd22e0b69afb0/src/Microsoft.TestPlatform.ObjectModel/Utilities/AssemblyLoadWorker.cs#L67.

Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: > It's funny because in the past you mentioned that we should enable it for all - even the inbox ones.

I do not think that I have said that. I [have said](https://github.com/dotnet/corefx/pull/32680#issuecomment-429478625) that if all inbox ones have it, CoreLib should have it too.

>  vstest for test assemblies:

These are not inbox assemblies.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: > I have said that if all inbox ones have it, CoreLib should have it too.

Thanks for clarifying.

> And I believe that this path is .NET Framework specific.

I know there was another code path somewhere for .NET Core but I can't find it right now...

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: If you think we should remove the property from all inbox assemblies we can do that in a follow-up PR.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 10]
```diff
@@ -62,6 +63,7 @@
     <_FullFrameworkReferenceAssemblyPaths>$(MSBuildThisFileDirectory)/Documentation</_FullFrameworkReferenceAssemblyPaths>
     <SkipCommonResourcesIncludes>true</SkipCommonResourcesIncludes>
     <DocumentationFile>$(OutputPath)$(MSBuildProjectName).xml</DocumentationFile>
+    <EnableAnalyzers>false</EnableAnalyzers>
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: @ViktorHofer Why did you disable the analyzers for CoreLib? They were enabled before your change.

Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Opened https://github.com/dotnet/runtime/issues/422 to have this tracked

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: https://github.com/dotnet/runtime/pull/424/files

**Review by safern — COMMENTED** [review]
LGTM
